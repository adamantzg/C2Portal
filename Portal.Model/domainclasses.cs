using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Linq;
using System.Web;

namespace Portal.Model
{
	public class Product
	{
		public int id { get; set; }
		public string code { get; set; }
		public string name { get; set; }
		public string description { get; set; }		
	}

	public class Customer
	{
		[Key]
		public string code { get; set; }
		public string name { get; set; }
		public string address1 { get; set; }
		public string address2 { get; set; }
		public string address3 { get; set; }
		public string address4 { get; set; }
		public string address5 { get; set; }
		public string address6 { get; set; }
		public string town_city { get; set; }
		public string county { get; set; }
		public string currency { get; set; }
		public string analysis_codes_1 { get; set; }
		public string invoice_customer { get; set; }
	}

	public class User
	{
		public int id { get; set; }
		public string name { get; set; }
		public string lastname { get; set; }
		public string username { get; set; }
		public string password { get; set; }
		public string email { get; set; }
		public string customer_code { get; set; }
		public string address { get; set; }
		public string phone { get; set; }
		public bool? isInternal { get; set; }
		public DateTime? lastLogin { get; set; }

		[NotMapped]
		public string token { get; set; }

		public virtual Customer Customer { get; set; }
		public virtual List<Role> Roles { get; set; }
		public virtual List<UserSession> Sessions { get; set; }

		[NotMapped]
		public List<Permission> Permissions { get; set; }

		public bool isAdmin => Roles?.Count(r => r.id == Role.Admin) > 0;
		public bool isBranchAdmin => Roles?.Count(r => r.id == Role.BranchAdmin) > 0;

		public bool HasPermission(PermissionId permId)
		{
			return Permissions?.Count(p => p.id == (int) permId) > 0;
		}
	}

	public class UserSession
	{
		public int id { get; set; }
		public int? user_id { get; set; }
		public DateTime? dateCreated { get; set; }
		public string token { get; set; }
		public string ip_addr { get; set; }
	}

	public class Role
	{
		public const int User = 1;
		public const int BranchAdmin = 2;
		public const int Admin = 3;

		public int id { get; set; }
		public string name { get; set; }

		public List<Permission> Permissions { get; set; }
	}
    public class Holiday
    {
        public int id { get; set; }
        public string name { get; set; }
        public DateTime date { get; set; }
    }

	[Table("permission")]
	public class Permission
	{
		public int id { get; set; }
		public string name { get; set; }
	}

	[Table("user_permission")]
	public class UserPermission
	{
		public int user_id { get; set; }
		public int permission_id { get; set; }
		public bool? grant { get; set; }
		public bool? deny { get; set; }

	}

	[Table("customer_permission")]
	public class CustomerPermission
	{
		public string customer_code { get; set; }
		public int permission_id { get; set; }
		public bool? grant { get; set; }
		public bool? deny { get; set; }

	}

	public enum PermissionId
	{
		ViewAccountDetails = 1,
		ViewHolidayAdministration = 2,
		ViewInvoiceHistory = 3,
		ViewOrderHistory = 4,
		ViewStockSearch = 5,
		ViewUserAdministration = 6,
		ViewRoleAdministration = 7
	}
}