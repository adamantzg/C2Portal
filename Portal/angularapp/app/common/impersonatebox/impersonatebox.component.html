
  <div class="panel">    
    <div class="form form-inline">
      <div class="form-group">
          <label *ngIf="showLabel">Impersonate: </label>
          <input [(ngModel)]="customer_code"
                    [typeahead]="customers"                    
                    [typeaheadOptionsLimit]="25"
                    (typeaheadLoading)="changeTypeaheadLoading($event)"
                    [typeaheadMinLength]="3"                    
                    typeaheadOptionField="combined_name"
                    placeholder="Type customer code"
                    class="form-control">          
          <button *ngIf="showButton" class="btn btn-primary m-l-xs" (click)="clicked()" [disabled]="!customer_code">Impersonate</button>
          <button class="btn btn-danger" (click)="clear(true)" *ngIf="customer">Clear</button>
          <label *ngIf="customer">{{customer.code}} {{customer.name}} {{customer.address6}}</label>
          <span class="alert alert-danger" *ngIf="showWarning">Customer with given code not found</span>
          <span *ngIf="autoCompleteLoading">Loading customers...</span>
      </div>      
    </div>
  </div>

