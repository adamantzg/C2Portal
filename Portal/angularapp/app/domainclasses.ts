
export class Product {
    id: number;
    code: string;
    name: string;
    description: string;
    combined_name: string;
}

export class Customer {
    code: string;
    name: string;
    address1: string;
    address2: string;
    address3: string;
    address4: string;
    address5: string;
    address6: string;
    town_city: string;
    county: string;
    analysis_codes1: string;
    currency: string;
    combined_name: string;
}

export class User {
    id: number;
    name: string;
    lastname: string;
    username: string;
    password: string;
    email: string;
    customer_code: string;
    address: string;
    phone: string;
    token: string;
    customer: Customer;
    roles: Role[];
    locked: boolean;
    isTopAdmin: boolean;
    isBranchAdmin: boolean;
    activated: boolean;
    isInternal: boolean;
    lastLogin: Date;
    role: string;
    role_id: number;
    permissions: Permission[];

    constructor() {
        this.customer = new Customer();
    }
}

export class Role {
    id: number;
    name: string;
    permissions: Permission[];
    // strPermissions: string[];

    constructor(id: number, name?: string, permissions?: Permission[]) {
        this.id = id;
        this.name = name;
        this.permissions = permissions;
    }
}


export class Invoice {
    customer: string;
    dated: Date | null;
    kind: string;
    reference: string;
    item_no: string;
    due_date: Date | null;
    amount: number | null;
    open_indicator: string;
    customer_order_num: string;
    selected: boolean;
    postcode: string;
}

export enum RoleId {
    User = 1,
    BranchAdmin = 2,
    Admin = 3
}

export class Order {
    order_no: string; // Sage order number
    customer_order_no: string; // Customer Order number
    date_entered: Date; // Date order received and input to Sage
    invoice_date: Date; // Currently will be to indicate "Delivered" flag. When there is an invoice date then the order is delivered
    date_required: Date;
    address1: string;
    address2: string;
    address3: string;
    address4: string;
    address5: string;
    address6: string;
    town_city: string;
    statusText: string;
    status: number;
    customer?: string;
    country: string;
    audits: OrderAudit[];
    productStockData: StockData[];
    details: OrderDetail[];
    planned: boolean;
}

export class OrderDetail {
    product: string;
    order_qty: number;
    val: number;
    order_line_no: number;
}

export  enum OrderStatus {
    Received = 5,
    InProgress = 7,
    Delivered = 8,
    Cancelled = 9
    // Processed = 'Processed',
    // Confirmed = 'Confirmed',
    // Delivered = 'Delivered',
    // Cancelled = 'Cancelled',
}
export class OrderAudit {
    audit_key: string; // need to join to the Order_no field of opheadm
    character_value: string; // limited to 6 to infer Despatch Note printed
    audit_date: Date; // is the date used for Order Processed date.
}

export class StockData {
    quantity: number;
    product: Product;
    ship_date: Date | null;
    discontinued: boolean;
    madeToOrder: boolean;
    expiresThisYear: boolean;
}

export class Holiday {
    name: string;
    date: Date;
}
export class HolidayExtracted {
    name: string;
    year: number;
    month: number;
    date: number;
    fullDate: Date;
}
export class Permission {
    id: number;
    name: string;
    selected: boolean;

    constructor(id?: number, name?: string, selected?: boolean) {
        this.id = id;
        this.name = name;
        this.selected = selected;
    }
}

export enum PermissionId {
    ViewAccountDetails = 1,
    ViewHolidayAdministration = 2,
    ViewInvoiceHistory = 3,
    ViewOrderHistory = 4,
    ViewStockSearch = 5,
    ViewUserAdministration = 6,
    ViewRoleAdministration = 7
}

export class ClearanceStock {
    product_code: string;
    long_description: string;
    range: string;
    exvat_price: number | null;
    price: number | null;
    freestock: number | null;
    discount: number;
}
