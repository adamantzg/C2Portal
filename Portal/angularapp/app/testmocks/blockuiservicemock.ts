import { EventEmitter } from '@angular/core';

export class BlockUIServiceMock {
    blockTicks = 0;
    unBlockTicks = 0;
    blockUIEvent: any = new EventEmitter<any>();

    startBlock() {
      this.blockTicks++;
    }

    stopBlock() {
      this.unBlockTicks++;
    }
}
