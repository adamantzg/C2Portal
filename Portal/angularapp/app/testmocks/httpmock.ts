import { of } from 'rxjs';

export class HttpMock {
    url: string;
    data: any;
    params: any;
    options: any;
    postResult: any;
    getResult: any;


    post<T>(url, data, params) {
      this.data = data;
      this.url = url;
      this.params = params;
      return of(this.postResult);
    }

    get<T>(url, options) {
        this.url = url;
        this.options = options;
        return of(this.getResult);
    }

  }
