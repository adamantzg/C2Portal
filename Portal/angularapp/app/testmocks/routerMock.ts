import { Component } from '@angular/core';

export class RouterMock {
    navigateUrls: string[];
    params: any;

    navigate(urls: string[], params?: any) {
        this.navigateUrls = urls;
        this.params = params;
    }
}

export class ActivatedRouteMock {

}


export class RouterLinkMock {

}

@Component({
    template: ''
  })
  export class DummyComponent {
  }
