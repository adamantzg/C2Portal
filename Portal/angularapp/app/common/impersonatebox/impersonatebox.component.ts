import { Component, OnInit, Input, Output, EventEmitter } from '@angular/core';
import { UserService } from '../../services/user.service';
import { Customer, RoleId } from '../../domainclasses';
import { Observable, Subject, Subscriber } from 'rxjs';

@Component({
  selector: 'app-impersonatebox',
  templateUrl: './impersonatebox.component.html',
  styleUrls: ['./impersonatebox.component.css']
})
export class ImpersonateboxComponent implements OnInit {

  constructor(private userService: UserService) { }

  customer_code: string;
  customer: Customer;

  /*@Input()
  public get Code() {
    return this.customer_code;
  }
  public set Code(value: string) {
    this.customer_code = value;
  }

  @Input()
  public get Customer() {
    return this.customerObj;
  }
  public set Customer(value: any)  {
    this.customerObj = value;
  }*/

  @Input()
  showLabel = false;

  @Input()
  showButton = true;

  /*@Output()
  buttonClicked = new EventEmitter();*/
  /*@Output()
  CodeChange = new EventEmitter<string>();*/
  @Output()
  CustomerChange = new EventEmitter<Customer>();
  showWarning = false;

  customers: Customer[] = [];
  autoCompleteLoading = false;

  ngOnInit() {
    const storage = localStorage.getItem('impersonated_customer');
    if (storage != null) {
      this.customer = JSON.parse(storage);
      this.CustomerChange.emit(this.customer);
      // this.CodeChange.emit(this.customerObj.code);
      this.customer_code = this.customer.code;
    } else {
      this.CustomerChange.emit(null);
    }
    this.customers = Observable.create((observer: any) => {
      // Runs on every search
      observer.next(this.customer_code);
    }).mergeMap((code: string) => this.userService.customerSearch(code, RoleId.User));

  }

  clicked() {
    // extract code
    const code = this.customer_code.split(' ')[0];
    this.userService.getCustomer(code).subscribe(c => {
      this.customer = c;
      this.CustomerChange.emit(this.customer);
      // this.CodeChange.emit(this.customer_code);
      if (c != null) {
        localStorage.setItem('impersonated_customer', JSON.stringify(c));
      } else {
        this.customer_code = '';
        this.showWarning = true;
        setTimeout(() => {
          this.showWarning = false;
        }, 1000 );
        localStorage.removeItem('impersonated_customer');
      }
    });
    // this.buttonClicked.emit();
  }

  clear() {
    this.customer_code = '';
    this.customer = null;
    localStorage.removeItem('impersonated_customer');
    this.CustomerChange.emit(null);
  }

  changeTypeaheadLoading(e: boolean) {
    this.autoCompleteLoading = e;
  }


}
