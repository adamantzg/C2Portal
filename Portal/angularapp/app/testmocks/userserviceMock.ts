import { User } from '../domainclasses';

export class UserServiceMock {
    CurrentUser: User;
}
