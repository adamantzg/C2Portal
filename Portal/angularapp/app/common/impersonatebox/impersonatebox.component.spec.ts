import { async, ComponentFixture, TestBed } from '@angular/core/testing';

import { ImpersonateboxComponent } from './impersonatebox.component';
import { FormsModule } from '@angular/forms';
import {RouterTestingModule} from '@angular/router/testing';
import {Router} from '@angular/router';

import { HttpClientModule, HttpClient } from '@angular/common/http';

import { routes } from '../../routestest';
import { UserService } from '../../services/user.service';
import { HttpService } from '../../services/http.service';
import { BlockUIService } from '../../../common_components/services/block-ui.service';
import { detectChanges } from '@angular/core/src/render3';
import { TypeaheadModule, ComponentLoaderFactory } from 'ngx-bootstrap';
import { HttpMock } from '../../testmocks/httpmock';
import { of } from 'rxjs';
import { Customer } from '../../domainclasses';

describe('ImpersonateboxComponent', () => {
  let component: ImpersonateboxComponent;
  let fixture: ComponentFixture<ImpersonateboxComponent>;
  let userService;

  setupTestBed({
    providers: [UserService, HttpService, {provide: HttpClient, useClass: HttpMock},  BlockUIService, ComponentLoaderFactory],
    imports: [RouterTestingModule.withRoutes(routes), FormsModule, TypeaheadModule.forRoot()],
    declarations: [ ImpersonateboxComponent ]

  });

  beforeEach(async(() => {
    fixture = TestBed.createComponent(ImpersonateboxComponent);
    component = fixture.componentInstance;
    userService = TestBed.get(UserService);
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(ImpersonateboxComponent);
    component = fixture.componentInstance;

  });

  test('should create', () => {
    expect(component).toBeTruthy();
    expect(fixture).toMatchSnapshot();
  });

  test('properties', () => {
    component.showLabel = true;
    fixture.detectChanges();

    expect(fixture.debugElement.nativeElement.querySelector('label')).toBeTruthy();
    const btn = fixture.debugElement.nativeElement.querySelector('.btn-primary');
    expect(btn).toBeTruthy();
    expect(btn.disabled).toBeTruthy();

    component.customer_code = 'a';
    fixture.detectChanges();
    expect(btn.disabled).toBeFalsy();

    const input = fixture.debugElement.nativeElement.querySelector('input');
    expect(input).toBeTruthy();
    const attrs = input.attributes;
    expect(attrs.getNamedItem('typeaheadoptionfield').value).toBe('combined_name');
    expect(attrs.getNamedItem('placeholder').value).toBe('Type customer code');

    expect(fixture.debugElement.nativeElement.querySelector('.alert-danger')).toBeNull();
    component.showWarning = true;
    fixture.detectChanges();
    expect(fixture.debugElement.nativeElement.querySelector('.alert-danger')).toBeTruthy();

  });

  test('buttonClicked', () => {
    component.customer_code = 'code name1 name2';
    let emitted;
    component.CustomerChange.subscribe(c => emitted = c );
    let returnedCust = {code: 'r1'};
    const fn = jest.spyOn(userService, 'getCustomer').mockImplementation(() => of(returnedCust));
    component.clicked();
    expect(fn.mock.calls[0][0]).toBe('code');
    expect(component.customer.code).toBe(returnedCust.code);
    expect(localStorage.getItem('impersonated_customer')).toEqual(JSON.stringify(returnedCust));
    expect(emitted.code).toBe(returnedCust.code);

    // No customer
    returnedCust = null;
    component.clicked();
    expect(component.customer_code).toBe('');
    expect(component.showWarning).toBeTruthy();
    expect(localStorage.getItem('impersonated_customer')).toBeNull();

  });

  test('onInit', () => {
    const cust = new Customer();
    cust.code = 'code';
    let emmited;
    localStorage.setItem('impersonated_customer', JSON.stringify(cust) );
    component.CustomerChange.subscribe(c => emmited = c);
    component.ngOnInit();
    expect(component.customer).toBeTruthy();
    expect(component.customer.code).toBe(cust.code);
    expect(component.customer_code).toBe(cust.code);
    expect(emmited).toBeTruthy();
    expect(emmited.code).toBe(cust.code);

  });

  test('clear', () => {
    component.customer_code = 'aaa';
    let emmited = {};
    localStorage.setItem('impersonated_customer', JSON.stringify({}) );
    component.CustomerChange.subscribe(c => emmited = c);
    component.customer = new Customer();
    component.clear();
    expect(component.customer_code).toBe('');
    expect(component.customer).toBeNull();
    expect(localStorage.getItem('impersonated_customer')).toBeNull();
    expect(emmited).toBeNull();
  });
});
