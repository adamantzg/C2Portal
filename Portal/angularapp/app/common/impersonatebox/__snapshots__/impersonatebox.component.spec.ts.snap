// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImpersonateboxComponent should create 1`] = `
<app-impersonatebox
  CustomerChange={[Function EventEmitter]}
  autoCompleteLoading="false"
  customers={[Function Array]}
  showButton={[Function Boolean]}
  showLabel="false"
  showWarning="false"
  userService={[Function UserService]}
>
  <div
    class="panel"
  >
    <div
      class="form form-inline"
    >
      <div
        class="form-group"
      >
        
        <input
          class="form-control"
          placeholder="Type customer code"
          typeaheadoptionfield="combined_name"
        />
        
        
        
        
        
      </div>
    </div>
  </div>
</app-impersonatebox>
`;
